File: client-sdk/go/types/inmsg.go
package types

import (
	"fmt"

	"github.com/oasisprotocol/oasis-core/go/common/cbor"
)

// LatestIncomingMessageVersion is the latest incoming message format version.
const LatestIncomingMessageVersion = 1

type IncomingMessageData struct {
	cbor.Versioned

	// UnverifiedTransaction is an embedded transaction (UnverifiedTransaction).
	// The transaction doesn't need to be from the same account that sent the message.
	UnverifiedTransaction *[]byte `json:"ut"`
}

func (d *IncomingMessageData) ValidateBasic() error {
	if d.V != LatestIncomingMessageVersion {
		return fmt.Errorf("incoming message data: unsupported version")
	}
	return nil
}

func NoopIncomingMessageData() *IncomingMessageData {
	return &IncomingMessageData{
		Versioned:             cbor.NewVersioned(LatestIncomingMessageVersion),
		UnverifiedTransaction: nil,
	}
}

File: runtime-sdk/src/types/in_msg.rs
use thiserror::Error;

/// The latest incoming message format version.
pub const LATEST_INCOMING_MESSAGE_VERSION: u16 = 1;

/// Error.
#[derive(Debug, Error)]
pub enum Error {
    #[error("unsupported version")]
    UnsupportedVersion,
    #[error("malformed incoming message data: {0}")]
    MalformedTransaction(anyhow::Error),
}

/// Roothash incoming message data.
#[derive(Clone, Debug, cbor::Encode, cbor::Decode)]
pub struct IncomingMessageData {
    #[cbor(rename = "v")]
    pub version: u16,
    /// An embedded transaction (UnverifiedTransaction in runtimes using this SDK).
    /// The transaction doesn't need to be from the same account that sent the message.
    pub tx: Option<Vec<u8>>,
}

impl IncomingMessageData {
    pub fn noop() -> Self {
        Self {
            version: LATEST_INCOMING_MESSAGE_VERSION,
            tx: None,
        }
    }

    pub fn validate_basic(&self) -> Result<(), Error> {
        if self.version != LATEST_INCOMING_MESSAGE_VERSION {
            return Err(Error::UnsupportedVersion);
        }
        Ok(())
    }
}
